ecl/eclscheduler/eclscheduler.cpp

```diff
@@ -180,6 +181,258 @@ version: "1.0"
+        Owned<IStringIterator> targetClusters = getTargetClusters("EclSchedulerProcess", processName);
+        ForEach(*targetClusters)
+        {
+            SCMStringBuffer targetCluster;
+            targetClusters->str(targetCluster);
+            Owned<EclScheduler> scheduler = new EclScheduler(this, targetCluster.str());
+            scheduler->start();
+            schedulers.append(*scheduler.getClear());
+        }
+        currentQueueNames.set(processName);
+#endif
+        if (schedulers.empty())
+            throw MakeStringException(0, "No clusters found to schedule for");
+
+        running = true;
+        LocalIAbortHandler abortHandler(*this);
+
+        while (running)
+        {
+            // wait for either a config change or an abort
+            waiter.wait();
+
+            if (running)
+            {
+                // Process a config change if one is pending
+                try
+                {
+                    CriticalBlock b(updateSchedulersCS);
+                    // onAbort could have triggered before or during the above switch, if so, we do no want to connect/block on new queue
+                    if (!running)
+                        break;
+
+                    bool newQueues = false;
+                    if (updatedQueueNames)
+                    {
+                        currentQueueNames.set(updatedQueueNames);
+                        updatedQueueNames.clear();
+                        newQueues = true;
+                    }
+
+                    if (newQueues)
+                        updateSchedulerQueues();
+                }
+                catch (IException *E)
+                {
+                    EXCLOG(E);
+                    releaseAtoms();
+                    ExitModuleObjects();
+                    _exit(2);
+                }
+                catch (...)
+                {
+                    IERRLOG("Unknown exception caught in eclscheduler processQueues - restarting");
+                    releaseAtoms();
+                    ExitModuleObjects();
+                    _exit(2);
+                }
+            }
+        }
+        DBGLOG("eclscheduler closing");
+        stop();
+    }
+
+private:
+    StringBuffer &getQueues(const IPropertyTree *config, StringBuffer &queueNameList)
+    {
+#ifdef _CONTAINERIZED
+        Owned<IPTreeIterator> queues = config->getElements("queues");
+        ForEach(*queues)
+        {
+            IPTree &queue = queues->query();
+            const char *qname = queue.queryProp("@name");
+            if (queueNameList.length())
+                queueNameList.append(",");
+            queueNameList.append(qname);
+        }
+#else
+        const char *processName = config->queryProp("@name");
+        queueNameList.append(processName);
+#endif
+        return queueNameList;
+    }
+
+    void updateSchedulerQueues()
+    {
+        std::unordered_set<std::string> updatedQueuesSet = splitBy(currentQueueNames.str(), ",");
```

Comment:
see other comment, retain a std::unordered_set<std::string> member, will simplify the code.